File: src/riot_pgen/gacc.cpp
//========================================================================================
// (C) (or copyright) 2024-2026. Triad National Security, LLC. All rights reserved.
//
// This program was produced under U.S. Government contract 89233218CNA000001 for Los
// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
// in the program are reserved by Triad National Security, LLC, and the U.S. Department
// of Energy/National Nuclear Security Administration. The Government is granted for
// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
// license in this material to reproduce, prepare derivative works, distribute copies to
// the public, perform publicly and display publicly, and to permit others to do so.
//========================================================================================
// This file was made in part with generative AI.

// gravitational acceleration of a sinusoidal density profile

#include <singularity-eos/eos/eos.hpp>

#include "riot_pgen/pgen.hpp"

namespace gacc {

using parthenon::ParArray1D;
using namespace RiotEOS;

//----------------------------------------------------------------------------------------
//! \fn  void gacc::ProblemGenerator
//! \brief
void ProblemGenerator(MeshBlock *pmb, ParameterInput *pin) {
  using parthenon::MakePackDescriptor;
  namespace ccbulk = cell_variables::cell_averaged::bulk;
  namespace ccmat = cell_variables::cell_averaged::mat;

  auto &rc = pmb->meshblock_data.Get();

  for (auto &var : rc->GetVariableVector()) {
    if (!var->IsAllocated()) pmb->AllocateSparse(var->label());
  }

  static auto desc =
      MakePackDescriptor<ccmat::rho, ccmat::volume_fraction, ccmat::internal_energy,
                         ccbulk::total_material_energy, ccbulk::momentum>(
          (pmb->resolved_packages).get());
  auto v = desc.GetPack(rc.get());

  // Adiabatic index (assuming single material)
  auto gravity_pkg = pmb->packages.Get("gravity");
  auto mat_pkg = pmb->packages.Get("materials");
  auto eos_vec = mat_pkg->Param<std::vector<RiotEOS::EOS>>("h.h.EOS");
  Real gm1 = eos_vec[0].GruneisenParamFromDensityTemperature(1., 1.);
  Real gamma = gm1 + 1.0;
  auto gdim = gravity_pkg->Param<int>("gravity_dim");
  auto gacc = gravity_pkg->Param<Real>("gravity_g");
  const int ndim = pmb->pmy_mesh->ndim;

  // Problem Parameters
  const Real P0 = pin->GetOrAddReal("problem", "P0", 2.5); // uniform pressure
  const Real vy = pin->GetOrAddReal("problem", "vy", 0.0); // initial velocity-y

  // Indexing
  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire);
  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);

  auto &coords = pmb->coords;

  pmb->par_for(
      "ProblemGenerator::gacc", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
      KOKKOS_LAMBDA(const int k, const int j, const int i) {
        Real x = coords.Xc<parthenon::X1DIR>(i);
        Real y = coords.Xc<parthenon::X2DIR>(j);
        Real z = coords.Xc<parthenon::X3DIR>(k);

        // sinusoidal density profile, evaluated pointwise at cell centroid
        Real rho = 1. + 0.1 * std::sin(2. * y * M_PI);
        v(0, ccmat::rho(0), k, j, i) = rho;
        v(0, ccmat::volume_fraction(0), k, j, i) = 1.0;

        // zero out velocities
        for (int dim = 0; dim < ndim; dim++) {
          v(0, ccbulk::momentum(dim), k, j, i) = 0.;
        }

        // initial velocity
        v(0, ccbulk::momentum(1), k, j, i) = rho * vy;

        // internal energy density
        Real u = P0 / (gamma - 1);

        // kinetic energy density
        Real ekin = 0.5 * rho * vy * vy;

        // total material energy
        v(0, ccbulk::total_material_energy(), k, j, i) = u + ekin;
      });

  return;
}

} // namespace gacc

File: inputs/gacc.py
#!/usr/bin/env python3
# ========================================================================================
#  (C) (or copyright) 2026. Triad National Security, LLC. All rights reserved.
#
#  This program was produced under U.S. Government contract 89233218CNA000001 for Los
#  Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
#  for the U.S. Department of Energy/National Nuclear Security Administration. All rights
#  in the program are reserved by Triad National Security, LLC, and the U.S. Department
#  of Energy/National Nuclear Security Administration. The Government is granted for
#  itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
#  license in this material to reproduce, prepare derivative works, distribute copies to
#  the public, perform publicly and display publicly, and to permit others to do so.
# ========================================================================================
# This file was made in part with generative AI.
# Set up sinusoidal density profile with uniform pressure and accelerate it
# through periodic boundaries using constant gravity
# gamma = 1.4, P0 = 2.5
#
# will travel 1 cm in sqrt(2/gacc) s.

import riot


def make_input():

    riot.input(
        "riot",
        problem="gacc",
    )

    riot.input(
        "parthenon/job",
        problem_id="gacceleration",
    )

    riot.input(
        "parthenon/output1",
        variables=[
            "c.c.mat.rho",
            "c.c.mat.volume_fraction",
            "c.c.bulk.velocity",
            "c.c.bulk.total_material_energy",
            "c.c.bulk.rho",
            "c.c.bulk.pressure",
            "c.c.bulk.bulk_modulus",
            "c.m.pressure",
        ],
        file_type="hdf5",
        dt=0.2,
    )

    riot.input(
        "parthenon/time",
        nlim=-1,
        tlim=0.4472135955,
        ncycle_out=20,
        integrator="rk2",
    )

    riot.input(
        "parthenon/mesh",
        nghost=2,
        refinement="none",
        numlevel=0,
        nx1=20,  # Number of zones in X1-direction
        x1min=0.0,  # minimum value of X1
        x1max=0.2,  # maximum value of X1
        ix1_bc="periodic",  # Inner-X1 boundary condition flag
        ox1_bc="periodic",  # Outer-X1 boundary condition flag
        nx2=100,  # Number of zones in X2-direction
        x2min=0.0,  # minimum value of X2
        x2max=1.0,  # maximum value of X2
        ix2_bc="periodic",  # Inner-X2 boundary condition flag
        ox2_bc="periodic",  # Outer-X2 boundary condition flag
        nx3=1,  # Number of zones in X3-direction
        x3min=-0.5,  # minimum value of X3
        x3max=0.5,  # maximum value of X3
        ix3_bc="periodic",  # Inner-X3 boundary condition flag
        ox3_bc="periodic",  # Outer-X3 boundary condition flag
    )

    riot.input(
        "parthenon/meshblock",
        nx1=20,
        nx2=100,
        nx3=1,
    )

    riot.input(
        "materials",
        sparse_dealloc=False,
    )

    riot.input(
        "material0",
        eos_type="IdealGas",
        Gamma=1.4,
        Cv=1.0e-3,
        max_mat_level=0,
        max_bnd_level=0,
    )

    riot.input(
        "physics",
        hydro=True,
        gravity=True,
    )

    riot.input(
        "hydro",
        recon="plm",
        cfl=0.8,
    )

    riot.input(
        "gravity",
        gravity_dim=1,
        gravity_g=-10.0,
    )


if __name__ == "__main__":
    make_input()
    riot.input.generate_input()
